src/schedule.ts

```diff
@@ -0,0 +1,72 @@
+import { Handle } from 'dojo-core/interfaces';
+import Scheduler from 'dojo-core/Scheduler';
+import { queueAnimationTask, QueueItem } from 'dojo-core/queue';
+
+/**
+ * Specifies a read or write operation
+ */
+enum Operation {
+	READ,
+	WRITE
+}
+
+/**
+ * A QueueItem representing a particular type of operation.
+ */
+interface DomQueueItem extends QueueItem {
+	operation: Operation;
+}
+
+/**
+ * An extension of Scheduler that allows read and write operations to be scheduled and dispatched in the proper order.
+ */
+class DomScheduler extends Scheduler {
+	/**
+	 * Overrides Scheduler#_dispatch to ensure that reads happen before writes.
+	 */
+	protected _dispatch(): void {
+		const reads: DomQueueItem[] = [];
+		const writes: DomQueueItem[] = [];
+		const queue = <DomQueueItem[]> this._queue;
+
+		for (const item of queue) {
+			(item.operation === Operation.READ ? reads : writes).push(item);
+		}
+
+		this._queue = reads.concat(writes);
+		super._dispatch();
+	}
+
+	/**
+	 * Schedules a read or write operation.
+	 */
+	scheduleOperation(callback: (...args: any[]) => void, operation: Operation): Handle {
+		return this._schedule({
+			isActive: true,
+			callback: callback,
+			operation: operation
+		});
+	}
+}
+
+const scheduler = new DomScheduler({ queueFunction: queueAnimationTask });
```

Comment:
This should be setting `deferWhileProcessing` to false to batch as many reads/writes together as possible - even if some are scheduled within read/write callbacks.

However, given how we're splitting the queue at the time `_dispatch` is called, I'm wondering how much complexity we'd have to add to it to accommodate this, or if there is a better approach to take.